core/chaincode/lifecycle/metadata_provider.go

```diff
@@ -119,37 +119,41 @@ func (mp *MetadataProvider) Metadata(channel string, ccName string, collections
 		spe = cauthdsl.MarshaledRejectAllPolicy
 	}
 
-	// process any existing collection endorsement policies
-	collectionPolicies := map[string][]byte{}
-	if ccInfo.Definition.Collections != nil {
-		for _, collectionName := range collections {
-			for _, conf := range ccInfo.Definition.Collections.Config {
-				staticCollConfig := conf.GetStaticCollectionConfig()
-				if staticCollConfig != nil && staticCollConfig.Name == collectionName {
-					if staticCollConfig.EndorsementPolicy != nil {
-						ep := protoutil.MarshalOrPanic(staticCollConfig.EndorsementPolicy)
-						cspe, err := mp.toSignaturePolicyEnvelope(channel, ep)
-						if err != nil {
-							logger.Errorf("could not convert collection policy for chaincode '%s' collection '%s' on channel '%s', err '%s'", ccName, collectionName, channel, err)
-							cspe = cauthdsl.MarshaledRejectAllPolicy
-						}
-						collectionPolicies[collectionName] = cspe
-					}
-				}
-			}
-		}
-	}
-
 	// report the sequence as the version to service discovery since
 	// the version is no longer required to change when updating any
 	// part of the chaincode definition
 	ccMetadata := &chaincode.Metadata{
 		Name:               ccName,
 		Version:            strconv.FormatInt(ccInfo.Definition.Sequence, 10),
 		Policy:             spe,
-		CollectionPolicies: collectionPolicies,
+		CollectionPolicies: map[string][]byte{},
 		CollectionsConfig:  ccInfo.Definition.Collections,
 	}
 
+	if ccInfo.Definition.Collections == nil {
+		return ccMetadata
+	}
+
+	// process any existing collection endorsement policies
+	for _, collectionName := range collections {
+		for _, conf := range ccInfo.Definition.Collections.Config {
+			staticCollConfig := conf.GetStaticCollectionConfig()
+			if staticCollConfig == nil {
+				continue
+			}
+			if staticCollConfig.Name == collectionName {
+				if staticCollConfig.EndorsementPolicy != nil {
+					ep := protoutil.MarshalOrPanic(staticCollConfig.EndorsementPolicy)
+					cspe, err := mp.toSignaturePolicyEnvelope(channel, ep)
+					if err != nil {
+						logger.Errorf("could not convert collection policy for chaincode '%s' collection '%s' on channel '%s', err '%s'", ccName, collectionName, channel, err)
+						cspe = cauthdsl.MarshaledRejectAllPolicy
```

Comment:
cspe = cauthdsl.MarshaledRejectAllPolicy

I missed this in the previous review. This wasn't introduced by you, but I don't think this should be here at all. 
If we failed converting this policy, then we should simply return nil and not use this RejectAllPolicy.

It makes sense to use RejectAllPolicy in validation, but this is a user facing API and we should return an error and not something that would make discovery analysis go haywire :-)

Reply:
Sounds good, will do for both the collection policy as well as chaincode policy.